File: programs/f0x01/src/instructions/commitment.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Mint, Token, TokenAccount, Transfer};
use crate::state::*;
use crate::error::*;
//create_commitment, claim_rewards

#[derive(Accounts)]
#[instruction(commitment_id: u64)]
pub struct CreateCommitment<'info> {
    #[account(
        init,
        payer = user,
        space = FocusCommitment::SPACE,
        seeds = [b"commitment", user.key().as_ref(), &commitment_id.to_le_bytes()],
        bump
    )]
    pub commitment: Account<'info, FocusCommitment>,
    
    #[account(
        mut,
        seeds = [b"user_profile", user.key().as_ref()],
        bump = user_profile.bump
    )]
    pub user_profile: Account<'info, UserProfile>,
    
    #[account(mut)]
    pub focus_program: Account<'info, FocusProgram>,
    
    #[account(mut)]
    pub user: Signer<'info>,
    
    #[account(
        mut,
        constraint = user_token_account.owner == user.key(),
        constraint = user_token_account.mint == focus_program.focus_token_mint
    )]
    pub user_token_account: Account<'info, TokenAccount>,
    
    #[account(
        init_if_needed,
        payer = user,
        seeds = [b"vault", user.key().as_ref(), &commitment_id.to_le_bytes()],
        bump,
        token::mint = token_mint,
        token::authority = vault_authority
    )]
    pub vault: Account<'info, TokenAccount>,
    
    /// CHECK: this is a PDA that acts as the vault authority and doesn't need type checking
    /// as it's not expected to be a deserialized account with specific data
    #[account(
        seeds = [b"vault_authority"],
        bump
    )]
    pub vault_authority: UncheckedAccount<'info>,
    
    #[account(address = focus_program.focus_token_mint)]
    pub token_mint: Account<'info, Mint>,
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

pub fn create_commitment(
    ctx: Context<CreateCommitment>,
    commitment_id: u64,
    amount: u64,
    sessions_per_day: u8,
    total_days: u8,
) -> Result<()> {
    require!(sessions_per_day > 0 && sessions_per_day <= 10, FocusError::InvalidSessionCount);
    require!(total_days > 0 && total_days <= 30, FocusError::InvalidDayCount);
    
    //transfer tokens to PDA vault
    let cpi_accounts = Transfer {
        from: ctx.accounts.user_token_account.to_account_info(),
        to: ctx.accounts.vault.to_account_info(),
        authority: ctx.accounts.user.to_account_info(),
    };
    
    let cpi_program = ctx.accounts.token_program.to_account_info();
    let cpi_ctx = CpiContext::new(cpi_program, cpi_accounts);
    token::transfer(cpi_ctx, amount)?;
    
    //initialize commitment state
    let commitment = &mut ctx.accounts.commitment;
    commitment.user = ctx.accounts.user.key();
    commitment.bump = ctx.bumps.commitment;
    commitment.commitment_id = commitment_id;
    commitment.amount_staked = amount;
    commitment.sessions_per_day = sessions_per_day;
    commitment.total_days = total_days;
    commitment.start_timestamp = Clock::get()?.unix_timestamp;
    commitment.days_completed = 0;
    commitment.is_active = true;
    commitment.last_session_timestamp = 0;
    commitment.sessions_completed_today = 0;
    
    //update program state
    let program = &mut ctx.accounts.focus_program;
    program.total_staked = program.total_staked.checked_add(amount).unwrap();
    
    Ok(())
}

#[derive(Accounts)]
pub struct ClaimRewards<'info> {
    #[account(
        mut,
        seeds = [b"commitment", user.key().as_ref(), &commitment.commitment_id.to_le_bytes()],
        bump = commitment.bump,
        constraint = commitment.user == user.key() @ FocusError::InvalidAuthority
    )]
    pub commitment: Account<'info, FocusCommitment>,
    
    #[account(
        mut,
        seeds = [b"user_profile", user.key().as_ref()],
        bump = user_profile.bump
    )]
    pub user_profile: Account<'info, UserProfile>,
    
    #[account(mut)]
    pub focus_program: Account<'info, FocusProgram>,
    
    #[account(mut)]
    pub user: Signer<'info>,
    
    #[account(
        mut,
        constraint = user_token_account.owner == user.key(),
        constraint = user_token_account.mint == focus_program.focus_token_mint
    )]
    pub user_token_account: Account<'info, TokenAccount>,
    
    #[account(
        mut,
        seeds = [b"vault", user.key().as_ref(), &commitment.commitment_id.to_le_bytes()],
        bump,
    )]
    pub vault: Account<'info, TokenAccount>,
    
    /// CHECK: this is a PDA that acts as the vault authority and doesn't need type checking
    /// as it's used only as a signer for token transfers
    #[account(
        seeds = [b"vault_authority"],
        bump
    )]
    pub vault_authority: UncheckedAccount<'info>,
    
    pub token_program: Program<'info, Token>,
}

pub fn claim_rewards(ctx: Context<ClaimRewards>) -> Result<()> {
    let commitment = &mut ctx.accounts.commitment;
    let user_profile = &mut ctx.accounts.user_profile;
    
    //verifyif the commitment has ended
    let current_timestamp = Clock::get()?.unix_timestamp;
    let day_in_seconds = 86400;
    let days_elapsed = ((current_timestamp - commitment.start_timestamp) / day_in_seconds) as u8;
    
    require!(days_elapsed >= commitment.total_days, FocusError::CommitmentNotEnded);
    require!(commitment.is_active, FocusError::CommitmentInactive);
    
    //calculate success rate
    let total_expected_sessions = commitment.sessions_per_day * commitment.total_days;
    let total_completed_sessions = user_profile.total_sessions_completed;
    let success_rate = (total_completed_sessions as f64) / (total_expected_sessions as f64);
    
    //calculate reward amount
    let program = &ctx.accounts.focus_program;
    let reward_amount = if success_rate >= 0.9 {
        //complete reward + bonus for 90%+ completion
        let base_reward = commitment.amount_staked;
        let bonus = (base_reward * program.reward_rate) / 100;
        base_reward.checked_add(bonus).unwrap()
    } else if success_rate >= 0.75 {
        //return original stake for 75%+ completion
        commitment.amount_staked
    } else {
        //partial refund for less than 75% completion
        (commitment.amount_staked * 75) / 100
    };
    
    //transfer reward tokens back to user
    let seeds = &[
        b"vault_authority".as_ref(),
        &[ctx.bumps.vault_authority],
    ];
    let signer = &[&seeds[..]];
    
    let cpi_accounts = Transfer {
        from: ctx.accounts.vault.to_account_info(),
        to: ctx.accounts.user_token_account.to_account_info(),
        authority: ctx.accounts.vault_authority.to_account_info(),
    };
    
    let cpi_program = ctx.accounts.token_program.to_account_info();
    let cpi_ctx = CpiContext::new_with_signer(cpi_program, cpi_accounts, signer);
    token::transfer(cpi_ctx, reward_amount)?;
    
    //update state
    commitment.is_active = false;
    user_profile.total_rewards_earned = user_profile.total_rewards_earned.checked_add(reward_amount).unwrap();
    
    //update the program state
    let program = &mut ctx.accounts.focus_program;
    program.total_staked = program.total_staked.checked_sub(commitment.amount_staked).unwrap();
    
    Ok(())
}

File: tests/commitment.ts
// OPTIMIZED VERSION OF COMMITMENT TESTS
// File: test/commitment.ts

import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { BN } from "@project-serum/anchor";
import { PublicKey, Keypair, SystemProgram } from "@solana/web3.js";
import {
  TOKEN_PROGRAM_ID,
  createMint,
  createAssociatedTokenAccount,
  mintTo,
  getAssociatedTokenAddress
} from "@solana/spl-token";
import { assert, expect } from "chai";
import { F0x01 } from "../target/types/f0x01";

// Shorter sleep function to avoid rate limiting
async function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe("F0x01 Commitment Tests", () => {
  // Configure the client
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const program = anchor.workspace.F0x01 as Program<F0x01>;
  const wallet = provider.wallet;
  
  // Global variables
  let focusProgramPda: PublicKey;
  let userKeypair: Keypair;
  let userProfilePda: PublicKey;
  let tokenMint: PublicKey;
  let userTokenAccount: PublicKey;
  
  // OPTIMIZATION: Use fewer commitment IDs to reduce transactions and rent costs
  const commitmentIds = {
    main: new anchor.BN(100),
    maxParams: new anchor.BN(101),
    duplicate: new anchor.BN(100), // Same as main to test duplication error
    wrongToken: new anchor.BN(102),
    simulation: new anchor.BN(103)
  };
  
  // OPTIMIZATION: Use smaller stake amount for tests that don't need large amounts
  const stakeAmount = new anchor.BN(10_000_000); // 10 tokens with 6 decimals (reduced from 100)
  
  // Store PDAs for reuse across tests
  let commitmentPda: PublicKey;
  let vaultPda: PublicKey;
  let vaultAuthorityPda: PublicKey;
  
  // Alternative commitment PDAs for other tests
  let maxParamsCommitmentPda: PublicKey;
  let maxParamsVaultPda: PublicKey;
  let wrongTokenCommitmentPda: PublicKey;
  let wrongTokenVaultPda: PublicKey;
  let simulationCommitmentPda: PublicKey;
  let simulationVaultPda: PublicKey;

  // OPTIMIZATION: Combined PDA lookup function
  function findCommitmentPdas(user: PublicKey, id: anchor.BN): [PublicKey, PublicKey] {
    const [cPda] = PublicKey.findProgramAddressSync(
      [
        Buffer.from("commitment"),
        user.toBuffer(),
        id.toArrayLike(Buffer, "le", 8)
      ],
      program.programId
    );

    const [vPda] = PublicKey.findProgramAddressSync(
      [
        Buffer.from("vault"),
        user.toBuffer(),
        id.toArrayLike(Buffer, "le", 8)
      ],
      program.programId
    );

    return [cPda, vPda];
  }

  // Reusable function to create commitment with better error handling
  async function createCommitment(
    id: anchor.BN,
    amount: anchor.BN,
    sessionsPerDay: number,
    totalDays: number,
    user: Keypair,
    userProfile: PublicKey,
    userToken: PublicKey
  ) {
    const [cPda, vPda] = findCommitmentPdas(user.publicKey, id);

    try {
      const tx = await program.methods
        .createCommitment(
          id,
          amount,
          sessionsPerDay,
          totalDays
        )
        .accountsStrict({
          commitment: cPda,
          userProfile: userProfile,
          focusProgram: focusProgramPda,
          user: user.publicKey,
          userTokenAccount: userToken,
          vault: vPda,
          vaultAuthority: vaultAuthorityPda,
          tokenMint: tokenMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          rent: anchor.web3.SYSVAR_RENT_PUBKEY,
        })
        .signers([user])
        .rpc({ commitment: 'confirmed' });
      
      // OPTIMIZATION: Shorter sleep
      await sleep(300);
      
      return { success: true, commitmentPda: cPda, vaultPda: vPda, tx };
    } catch (error) {
      return { success: false, error, commitmentPda: cPda, vaultPda: vPda };
    }
  }

  before(async () => {
    // console.log("Setting up optimized test environment...");

    // Find the focus_program PDA
    [focusProgramPda] = PublicKey.findProgramAddressSync(
      [Buffer.from("focus_program")],
      program.programId
    );

    // OPTIMIZATION: Reuse the same keypair for all tests to save transaction costs
    userKeypair = Keypair.generate();

    // OPTIMIZATION: Fund the user with much less SOL - just enough for account rents
    // Instead of 2 SOL, use 0.05 SOL (50,000,000 lamports)
    const fundTx = new anchor.web3.Transaction().add(
      SystemProgram.transfer({
        fromPubkey: wallet.publicKey,
        toPubkey: userKeypair.publicKey,
        lamports: 50000000, // 0.05 SOL
      })
    );
    await provider.sendAndConfirm(fundTx);
    await sleep(500);

    // Find the user profile PDA
    [userProfilePda] = PublicKey.findProgramAddressSync(
      [Buffer.from("user_profile"), userKeypair.publicKey.toBuffer()],
      program.programId
    );

    // Find the vault authority PDA
    [vaultAuthorityPda] = PublicKey.findProgramAddressSync(
      [Buffer.from("vault_authority")],
      program.programId
    );

    // Find commitment PDAs for all test cases upfront to avoid redundant calculations
    [commitmentPda, vaultPda] = findCommitmentPdas(userKeypair.publicKey, commitmentIds.main);
    [maxParamsCommitmentPda, maxParamsVaultPda] = findCommitmentPdas(userKeypair.publicKey, commitmentIds.maxParams);
    [wrongTokenCommitmentPda, wrongTokenVaultPda] = findCommitmentPdas(userKeypair.publicKey, commitmentIds.wrongToken);
    [simulationCommitmentPda, simulationVaultPda] = findCommitmentPdas(userKeypair.publicKey, commitmentIds.simulation);

    // OPTIMIZATION: Check if program is already initialized before doing token setup
    try {
      const programAccount = await program.account.focusProgram.fetch(focusProgramPda);
      tokenMint = programAccount.focusTokenMint;
      // console.log("Using existing program and token mint");
      await sleep(200);
    } catch (error) {
      console.error("Program not initialized. Please run initialization test first.");
      throw new Error("Program must be initialized before running commitment tests");
    }

    // OPTIMIZATION: Combined token account setup
    try {
      userTokenAccount = await getAssociatedTokenAddress(tokenMint, userKeypair.publicKey);

      // Check if token account exists and create if needed
      try {
        await provider.connection.getTokenAccountBalance(userTokenAccount);
        // console.log("Using existing token account");
      } catch (error) {
        // Create token account if it doesn't exist
        // console.log("Creating token account for user");
        userTokenAccount = await createAssociatedTokenAccount(
          provider.connection,
          wallet.payer,
          tokenMint,
          userKeypair.publicKey,
          { commitment: 'confirmed' }
        );
        await sleep(300);
      }

      // OPTIMIZATION: Mint fewer tokens, just enough for tests
      // Mint 50 tokens (with 6 decimals) instead of 1000
      const tokensNeeded = 50_000_000;
      await mintTo(
        provider.connection,
        wallet.payer,
        tokenMint,
        userTokenAccount,
        wallet.publicKey,
        tokensNeeded,
        [],
        { commitment: 'confirmed' }
      );
      await sleep(300);

    } catch (error) {
      console.error("Error setting up token accounts:", error);
      throw error;
    }

    // Create user profile if it doesn't exist
    try {
      await program.account.userProfile.fetch(userProfilePda);
      console.log("Using existing user profile");
    } catch (error) {
      // console.log("Creating user profile");
      await program.methods
        .createUserProfile()
        .accountsStrict({
          userProfile: userProfilePda,
          user: userKeypair.publicKey,
          focusProgram: focusProgramPda,
          systemProgram: SystemProgram.programId,
        })
        .signers([userKeypair])
        .rpc({ commitment: 'confirmed' });
      
      await sleep(300);
    }
  });

  // Test 1: Create commitment with minimum valid parameters
  it("Creates commitment with minimum valid parameters", async () => {
    const result = await createCommitment(
      commitmentIds.main,
      stakeAmount,
      1, // minSessionsPerDay
      1, // minTotalDays
      userKeypair,
      userProfilePda,
      userTokenAccount
    );

    if (!result.success) {
      console.error("Failed to create commitment:", result.error);
      throw result.error;
    }

    // Verify the commitment creation
    const commitment = await program.account.focusCommitment.fetch(commitmentPda);
    expect(commitment.user.toString()).to.equal(userKeypair.publicKey.toString());
    expect(commitment.sessionsPerDay).to.equal(1);
    expect(commitment.totalDays).to.equal(1);
    expect(commitment.isActive).to.be.true;

    // Verify vault received the tokens
    const vaultBalance = await provider.connection.getTokenAccountBalance(vaultPda);
    expect(vaultBalance.value.amount).to.equal(stakeAmount.toString());
    
    await sleep(300);
  });

  // Test 2: Create commitment with maximum valid parameters
  it("Creates commitment with maximum valid parameters", async () => {
    const result = await createCommitment(
      commitmentIds.maxParams,
      stakeAmount,
      10, // maxSessionsPerDay
      30, // maxTotalDays
      userKeypair,
      userProfilePda,
      userTokenAccount
    );

    if (!result.success) {
      console.error("Failed to create max commitment:", result.error);
      throw result.error;
    }

    // Verify the commitment was created correctly
    const commitment = await program.account.focusCommitment.fetch(maxParamsCommitmentPda);
    expect(commitment.user.toString()).to.equal(userKeypair.publicKey.toString());
    expect(commitment.sessionsPerDay).to.equal(10);
    expect(commitment.totalDays).to.equal(30);
    expect(commitment.isActive).to.be.true;
    
    await sleep(300);
  });

  // Test 3: Duplicate commitment ID (using the same ID as the first test)
  it("Fails to create commitment with duplicate ID", async () => {
    try {
      // Try to create commitment with same ID as the first test
      await program.methods
        .createCommitment(
          commitmentIds.duplicate, // This is the same as commitmentIds.main
          stakeAmount,
          2,
          2
        )
        .accountsStrict({
          commitment: commitmentPda,
          userProfile: userProfilePda,
          focusProgram: focusProgramPda,
          user: userKeypair.publicKey,
          userTokenAccount: userTokenAccount,
          vault: vaultPda,
          vaultAuthority: vaultAuthorityPda,
          tokenMint: tokenMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          rent: anchor.web3.SYSVAR_RENT_PUBKEY,
        })
        .signers([userKeypair])
        .rpc({ commitment: 'confirmed' });
      
      assert.fail("Should have failed due to duplicate commitment ID");
    } catch (error) {
      // Expected to fail because account already exists
      expect(error.message).to.include("Error");
    }
    
    await sleep(300);
  });

  // Test 4: Invalid token account (simplified to use less SOL)
  it("Fails to create commitment with invalid token account", async () => {
    // OPTIMIZATION: Simulate failure by using incorrect accounts instead of creating new token mints
    try {
      await program.methods
        .createCommitment(
          commitmentIds.wrongToken,
          stakeAmount,
          2,
          2
        )
        .accountsStrict({
          commitment: wrongTokenCommitmentPda,
          userProfile: userProfilePda,
          focusProgram: focusProgramPda,
          user: userKeypair.publicKey,
          userTokenAccount: focusProgramPda, // Using program PDA as token account (which will fail)
          vault: wrongTokenVaultPda,
          vaultAuthority: vaultAuthorityPda,
          tokenMint: tokenMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          rent: anchor.web3.SYSVAR_RENT_PUBKEY,
        })
        .signers([userKeypair])
        .rpc({ commitment: 'confirmed' });
      
      assert.fail("Should have failed due to invalid token account");
    } catch (error) {
      // Expected to fail due to constraint violation
      expect(error.message).to.include("Error");
    }
    
    await sleep(300);
  });

  // Test 5: Simulation test (without on-chain transactions)
  it("Simulates completing a commitment and claiming rewards", async () => {
    // Create one actual commitment for the simulation
    const result = await createCommitment(
      commitmentIds.simulation,
      stakeAmount,
      2, // sessionsPerDay
      1, // totalDays
      userKeypair,
      userProfilePda,
      userTokenAccount
    );
    
    if (!result.success) {
      console.error("Failed to create simulation commitment:", result.error);
      throw result.error;
    }
    
    // Get the commitment data
    const commitment = await program.account.focusCommitment.fetch(simulationCommitmentPda);
    await sleep(200);
    
    // Get the program details
    const programData = await program.account.focusProgram.fetch(focusProgramPda);
    const rewardRate = programData.rewardRate.toNumber();
    await sleep(200);
    
    // Total expected sessions
    const totalExpectedSessions = commitment.sessionsPerDay * commitment.totalDays;
    
    // Define scenarios without creating additional transactions
    const scenarios = [
      { 
        name: "High Completion (90%+)", 
        completedSessions: totalExpectedSessions, // 100% completion
        description: "User completes 90% or more of required sessions"
      },
      { 
        name: "Medium Completion (75-89%)", 
        completedSessions: Math.floor(totalExpectedSessions * 0.8), // 80% completion
        description: "User completes between 75% and 89% of required sessions"
      },
      { 
        name: "Low Completion (<75%)", 
        completedSessions: Math.floor(totalExpectedSessions * 0.5), // 50% completion
        description: "User completes less than 75% of required sessions"
      }
    ];

    // Calculate reward scenarios without creating transactions
    for (const scenario of scenarios) {
      const successRate = scenario.completedSessions / totalExpectedSessions;
      
      // Calculate rewards based on completion thresholds
      let rewardAmount;
      
      if (successRate >= 0.9) {
        // Full stake back plus bonus
        const baseStake = commitment.amountStaked.toNumber();
        const bonus = (baseStake * rewardRate) / 100;
        rewardAmount = baseStake + bonus;
      } else if (successRate >= 0.75) {
        // Return original stake only
        rewardAmount = commitment.amountStaked.toNumber();
      } else {
        // Partial refund (75% of stake)
        rewardAmount = Math.floor(commitment.amountStaked.toNumber() * 0.75);
      }
      
      // Just verify calculation works, no need to check values
      expect(rewardAmount).to.be.a('number');
    }
  });
});